File: src/main/java/pl/training/module3/domain/DeliveryQuote.java
package pl.training.module3.domain;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

public record DeliveryQuote(
        String customerEmail,
        ShippingMethod method,
        Parcel parcel,
        BigDecimal price) {
    public DeliveryQuote {
        Objects.requireNonNull(customerEmail, "customerEmail");
        Objects.requireNonNull(method, "method");
        Objects.requireNonNull(parcel, "parcel");
        Objects.requireNonNull(price, "price");

        if (customerEmail.isBlank()) {
            throw new IllegalArgumentException("Customer email must not be blank");
        }
        if (price.signum() < 0) {
            throw new IllegalArgumentException("Price must not be negative");
        }

        price = price.setScale(2, RoundingMode.UNNECESSARY);
    }
}

File: csharp/src/Training.Module3/Domain/DeliveryQuote.cs
namespace Training.Module3.Domain;

public sealed record DeliveryQuote
{
    public DeliveryQuote(
        string customerEmail,
        ShippingMethod method,
        Parcel parcel,
        decimal price)
    {
        ArgumentNullException.ThrowIfNull(customerEmail);
        ArgumentNullException.ThrowIfNull(parcel);

        if (string.IsNullOrWhiteSpace(customerEmail))
        {
            throw new ArgumentException("Customer email must not be blank", nameof(customerEmail));
        }
        if (price < 0m)
        {
            throw new ArgumentException("Price must not be negative", nameof(price));
        }

        CustomerEmail = customerEmail;
        Method = method;
        Parcel = parcel;
        Price = ScaleTwoWithoutRounding(price);
    }

    public string CustomerEmail { get; }

    public ShippingMethod Method { get; }

    public Parcel Parcel { get; }

    public decimal Price { get; }

    // Equivalent of BigDecimal.setScale(2, RoundingMode.UNNECESSARY):
    // widens the scale to two digits and refuses to silently drop precision.
    private static decimal ScaleTwoWithoutRounding(decimal price)
    {
        decimal scaled = decimal.Round(price * 1.00m, 2);
        if (scaled != price)
        {
            throw new ArithmeticException("Rounding necessary");
        }
        return scaled;
    }
}
